personhood/contract_popparty.go

```diff
@@ -0,0 +1,492 @@
+func PopPartyFinalize(cl *byzcoin.Client, popIID byzcoin.InstanceID, atts Attendees, signers ...darc.Signer) (err error) {
+	var sigStrs []string
+	for _, sig := range signers {
+		sigStrs = append(sigStrs, sig.Identity().String())
+	}
+	signerCtrs, err := cl.GetSignerCounters(sigStrs...)
+	if err != nil {
+		return
+	}
+
+	attBuff, err := protobuf.Encode(&atts)
+	if err != nil {
+		return
+	}
+	ctx := byzcoin.ClientTransaction{
+		Instructions: byzcoin.Instructions{byzcoin.Instruction{
+			InstanceID: popIID,
+			Invoke: &byzcoin.Invoke{
+				ContractID: ContractPopPartyID,
+				Command:    "finalize",
+				Args: byzcoin.Arguments{
+					{
+						Name:  "attendees",
+						Value: attBuff,
+					},
+				},
+			},
+			SignerCounter: []uint64{signerCtrs.Counters[0] + 1},
+		}},
+	}
+	err = ctx.FillSignersAndSignWith(signers...)
+	if err != nil {
+		return
+	}
+	_, err = cl.AddTransactionAndWait(ctx, 5)
+	return
+}
+
+// PopPartyMine collects the reward. If 'atts' is nil, the party will be fetched from
+// byzcoin. One of 'coinIID' or 'd' must be set to either mine the coin, or spawn
+// a new darc and coin.
+func PopPartyMine(cl *byzcoin.Client, popIID byzcoin.InstanceID, kp key.Pair,
+	atts *Attendees, coinIID *byzcoin.InstanceID, d *darc.Darc) (err error) {
+	if atts == nil {
+		popProof, err := cl.GetProof(popIID.Slice())
+		if err != nil {
+			return err
+		}
+		_, value, cID, _, err := popProof.Proof.KeyValue()
+		if err != nil {
+			return err
+		}
+		if cID != ContractPopPartyID {
+			return errors.New("given popIID is not of contract-type PopParty")
+		}
+		var pop PopPartyStruct
+		err = protobuf.DecodeWithConstructors(value, &pop, network.DefaultConstructors(cothority.Suite))
+		if err != nil {
+			return err
+		}
+		atts = &pop.Attendees
+	}
+	var mine = -1
+	for i, p := range atts.Keys {
+		if p.Equal(kp.Public) {
+			mine = i
+			break
+		}
+	}
+	if mine == -1 {
+		return errors.New("didn't find public key of keypair in attendees")
+	}
+
+	lrs := anon.Sign(&suiteBlake2s{}, []byte("mine"), atts.Keys, popIID[:], mine, kp.Private)
+	args := byzcoin.Arguments{{
+		Name:  "lrs",
+		Value: lrs,
+	}}
+	if coinIID == nil {
+		darcBuf, err := d.ToProto()
+		if err != nil {
+			return err
+		}
+
+		args = append(args, byzcoin.Argument{
+			Name:  "newDarc",
+			Value: darcBuf,
+		})
+	} else {
+		args = append(args, byzcoin.Argument{
+			Name:  "coinID",
+			Value: coinIID.Slice(),
+		})
+	}
+
+	ctx := byzcoin.ClientTransaction{
+		Instructions: byzcoin.Instructions{byzcoin.Instruction{
+			InstanceID: popIID,
+			Invoke: &byzcoin.Invoke{
+				ContractID: ContractPopPartyID,
+				Command:    "mine",
+				Args:       args,
+			},
+		}},
+	}
+	_, err = cl.AddTransactionAndWait(ctx, 5)
+	return
+}
+
+// PopPartyMineDarcToCoin calculates the coin given a darc and returns the coin instance.
+func PopPartyMineDarcToCoin(cl *byzcoin.Client, d *darc.Darc) (coinIID byzcoin.InstanceID, coin byzcoin.Coin, err error) {
+	h := sha256.New()
```

Comment:
This code is duplicated up a few hundred lines in this file.

Reply:
Yes, but no. This method is to be used from a client, the `sha256("coin" | darc.baseID)` further up doesn't have a `byzcoin.Client`. 

Perhaps `PopPartyMineDarcToCoin` and `PopPartyMine` should go into `api.go`?